[SPARK-43144] Scala Client DataStreamReader table() API

WweiL opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Add the table() API for scala client.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Continuation of SS Connect development.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Unit test

I also changed `ProtoToParsedPlanTestSuite` a little to remove the memory addresses, before the change the test for streaming table would fail with:
```
- streaming_table_API_with_options *** FAILED *** (8 milliseconds)
[info]   Expected and actual plans do not match:
[info]   
[info]   === Expected Plan ===
[info]   SubqueryAlias primary.tempdb.myStreamingTable
[info]   +- StreamingRelationV2 primary.tempdb.myStreamingTable, org.apache.spark.sql.connector.catalog.InMemoryTable@752725d9, [p1=v1, p2=v2], [id#0L], org.apache.spark.sql.connector.catalog.InMemoryCatalog@347d8e2a, tempdb.myStreamingTable
[info]   
[info]   
[info]   === Actual Plan ===
[info]   SubqueryAlias primary.tempdb.myStreamingTable
[info]   +- StreamingRelationV2 primary.tempdb.myStreamingTable, org.apache.spark.sql.connector.catalog.InMemoryTable@a88a5db, [p1=v1, p2=v2], [id#0L], org.apache.spark.sql.connector.catalog.InMemoryCatalog@2c6b362e, tempdb.myStreamingTable
```
Because the memory address (`InMemoryTable@752725d9`) is different every time it runs. I removed these in the test suite.


And verified that memory addresses doesn't exist in existing explain files:
```
wei.liu:~/oss-spark$ cat connector/connect/common/src/test/resources/query-tests/explain-results/* | grep @
wei.liu:~/oss-spark$ 
```

WweiL committed:
done?

WweiL committed:
wip

WweiL committed:
done

rangadi commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:
```diff
@@ -78,4 +78,44 @@ class StreamingQuerySuite extends RemoteSparkSession with SQLHelper {
+            q1.processAllAvailable()
+            q2.processAllAvailable()
+            eventually(timeout(10.seconds)) {
+              assert(q2.recentProgress.nonEmpty) // Query made progress.
```
This does not mean if would have processed any data, right? We could move `assert(spark.table("my_sink").count() > 0)` also into this.

rangadi commented on connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala:
```diff
@@ -217,6 +217,20 @@ final class DataStreamReader private[sql] (sparkSession: SparkSession) extends L
+  def table(tableName: String): DataFrame = {
+    require(tableName != null, "The table name can't be null")
+    sparkSession.newDataFrame { builder =>
+      builder.getReadBuilder.setIsStreaming(true).getNamedTableBuilder
```
Need set options.

rangadi replied:
Also, it will more readable to split this into more than one line.

WweiL replied:
Sure I'll make it multiple lines. For options I thought they are already set in the option method? 
https://github.com/apache/spark/blob/ad58f4c99eb9957d8f6bf55aa8dc7f7d48692f31/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala#L82

Here we don't have the `extraOptions` like what did in `DataFrameReader`

amaliujia replied:
Is there a way to also test the options? Maybe `spark.readStream().option().table()`?

WweiL replied:
On second thought I think we do need to set options. I've updated the code. I had a hard time finding an option for table though..

amaliujia replied:
You can refer to https://github.com/apache/spark/pull/40498 to how that is tested.

WweiL replied:
Cool thank you!

WweiL committed:
address comments

WweiL committed:
add options for table

WweiL committed:
option test

WweiL committed:
generate golden file

WweiL committed:
minor

amaliujia commented:
You probably also need to generate the golden file for `ProtoToParsedPlanTestSuite`. There is instructions documented in that suite.

WweiL commented:
> You probably also need to generate the golden file for `ProtoToParsedPlanTestSuite`. There is instructions documented in that suite.

Ah I see there is also a bin file. I did run this but my local dev env is broken so I ran this on devbox... and scp the file back.. will also scp that one

WweiL committed:
protoToParsedPlanTestSuite

amaliujia reviewed (approved):
LGTM

WweiL commented:
@HyukjinKwon Can you merge this when you get a chance? Thank you!

WweiL committed:
protoToParsedPlan remove MemoryAddress, now test should pass

WweiL commented:
@amaliujia @HyukjinKwon 
I also changed `ProtoToParsedPlanTestSuite` a little to remove the memory addresses, before the change the test for streaming table only would fail with:
```
- streaming_table_API_with_options *** FAILED *** (8 milliseconds)
[info]   Expected and actual plans do not match:
[info]   
[info]   === Expected Plan ===
[info]   SubqueryAlias primary.tempdb.myStreamingTable
[info]   +- StreamingRelationV2 primary.tempdb.myStreamingTable, org.apache.spark.sql.connector.catalog.InMemoryTable@752725d9, [p1=v1, p2=v2], [id#0L], org.apache.spark.sql.connector.catalog.InMemoryCatalog@347d8e2a, tempdb.myStreamingTable
[info]   
[info]   
[info]   === Actual Plan ===
[info]   SubqueryAlias primary.tempdb.myStreamingTable
[info]   +- StreamingRelationV2 primary.tempdb.myStreamingTable, org.apache.spark.sql.connector.catalog.InMemoryTable@a88a5db, [p1=v1, p2=v2], [id#0L], org.apache.spark.sql.connector.catalog.InMemoryCatalog@2c6b362e, tempdb.myStreamingTable
```
Because the memory address is different every time it runs. I removed these in the test suite.


And verified that memory addresses doesn't exist in existing explain files:
```
wei.liu:~/oss-spark$ cat connector/connect/common/src/test/resources/query-tests/explain-results/* | grep @
wei.liu:~/oss-spark$ 
```

PTAL, thank you!

WweiL committed:
fmt

zhengruifeng commented:
merged to master
